nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureBlobStorage.java

```diff
@@ -106,10 +110,15 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
             final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
             session.getProvenanceReporter().send(flowFile, blob.getSnapshotQualifiedUri().toString(), transferMillis);
 
-        } catch (IllegalArgumentException | URISyntaxException | StorageException e) {
-            getLogger().error("Failed to put Azure blob {}", new Object[]{blobPath}, e);
-            flowFile = session.penalize(flowFile);
-            session.transfer(flowFile, REL_FAILURE);
+        } catch (IllegalArgumentException | URISyntaxException | StorageException | ProcessException e) {
+            if (e instanceof ProcessException && storedException.get() == null) {
+                throw (ProcessException) e;
+            } else {
+                Exception failureException = Optional.ofNullable(storedException.get()).map(x -> x).orElse(e);
```

Comment:
@jtstorck I don't think you need the .map() here. It looks like orElse() will already give you the value if the optional is absent.